### PR TITLE
Backup restoring

### PR DESCRIPTION
**Pull request by djklim87**

### CODE REVIEW

File: docker-entrypoint.sh [hunk 1]
```diff
@@ -205,6 +210,36 @@ _main() {
   fi
 
   _replace_conf_from_env
+
+  BACKUP_INIT_FOLDER="/docker-entrypoint-initdb.d"
+
+  if [[ $(ls -1 $BACKUP_INIT_FOLDER | wc -l) -eq 4 ]]; then
```
Comment (sanikolaev) [inline docker-entrypoint.sh:216]: What's the idea of comparing with 4 here? Please add a comment.

File: docker-entrypoint.sh [hunk 2]
```diff
@@ -205,6 +210,36 @@ _main() {
...
+
+    if [ $(ls -la /usr/bin/manticore-executor | awk '{print $5}') -eq 0 ]; then
```
Comment (sanikolaev) [inline docker-entrypoint.sh:218]: Please use a simpler command - https://www.gnu.org/software/bash/manual/html_node/Bash-Conditional-Expressions.html

File: docker-entrypoint.sh [hunk 3]
```diff
@@ -205,6 +210,36 @@ _main() {
...
+        echo -e "${RED}Can't run manticore-backup. EXTRA packages wasn't installed${NC}"
```
Comment (sanikolaev) [inline docker-entrypoint.sh:219]:
```suggestion
        echo -e "${RED}Can't run manticore-backup. Use env. var. EXTRA=1 to install the missing packages.${NC}"
```

File: docker-entrypoint.sh [hunk 4]
```diff
@@ -205,6 +210,36 @@ _main() {
...
+        exit 1
+    fi
+
+    [[ $(which manticore-backup) ]] || \
+      { echo -e "${RED}Manticore backup doesn't installed${NC}"; exit 1; }
```
Comment (sanikolaev) [inline docker-entrypoint.sh:224]:
```suggestion
      { echo -e "${RED}manticore-backup isn't installed.${NC}"; exit 1; }
```

File: docker-entrypoint.sh [hunk 5]
```diff
@@ -205,6 +210,36 @@ _main() {
...
+
+    if [ ! -f "${BACKUP_INIT_FOLDER}/versions.json" ]; then
+        echo -e "${RED}Dump is corrupted${NC}"
```
Comment (sanikolaev) [inline docker-entrypoint.sh:227]:
```suggestion
        echo -e "${RED}Dump is corrupted.${NC}"
```

File: docker-entrypoint.sh [hunk 6]
```diff
@@ -205,6 +210,36 @@ _main() {
...
+    fi
+
+    find ${BACKUP_INIT_FOLDER}/config -type f -exec sh -c 'rm -f "${1#/docker-entrypoint-initdb.d/config}"' sh {} \;
+    find ${BACKUP_INIT_FOLDER}/state -type f -exec sh -c 'rm -f "${1#/docker-entrypoint-initdb.d/state}"' sh {} \;
+
+    manticore-backup --version
+    manticore-backup --force --backup-dir='/' --restore='docker-entrypoint-initdb.d'
+
+    if [ -z "$START_AFTER_RESTORE" ]; then
+        echo -e "${GREEN}Dump successfully restored.${NC} Run container again without mount anything to docker-entrypoint-initdb.d"
```
Comment (sanikolaev) [inline docker-entrypoint.sh:238]:
```suggestion
        echo -e "${GREEN}Dump successfully restored.${NC} Run the container again without mounting anything to docker-entrypoint-initdb.d"        
```

File: Dockerfile [hunk 7]
```diff
@@ -92,12 +92,14 @@ RUN if [ "$TARGETPLATFORM" = "linux/arm64" ] ; then export ARCH="arm"; else expo
 RUN if [ -d "/packages/" ]; then apt -y install /packages/*deb; fi \
     && mkdir -p /var/run/manticore \
     && mkdir -p /var/lib/manticore/replication \
+    && mkdir /docker-entrypoint-initdb.d \
```
Comment (sanikolaev) [inline Dockerfile:95]: Doesn't it get created automatically if the user maps an outside dir to it?

File: README.md [hunk 8]
```diff
@@ -312,6 +312,52 @@ There are several methods to build plain tables from your custom configuration f
 4) **Combining cron-based and startup table rebuilding:**  
    To combine cron-based indexing with the indexing of desired tables on startup, use this format: `CREATE_PLAIN_TABLES=tbl:* * * * *;tbl2:*/5 2 * * *;deltaTable;tbl3`.
 
+# Backup And Restore
```
Comment (sanikolaev) [inline README.md:315]:
```suggestion
# Backup and restore
```

File: README.md [hunk 9]
```diff
@@ -312,6 +312,52 @@ There are several methods to build plain tables from your custom configuration f
...
+
+```bash
+$ ls /tmp/ | grep backup-*
+backup-20230509133521
+```
+Inside this folder, you will find your backup.
+
+### Restore full dump
+
+We have automated the backup restoration process on startup, following the typical approach for Docker images of MySQL or other common databases.
```
Comment (sanikolaev) [inline README.md:337]:
```suggestion
```

File: README.md [hunk 10]
```diff
@@ -312,6 +312,52 @@ There are several methods to build plain tables from your custom configuration f
...
+
+To restore your full backup, you need to mount your backup to the `/docker-entrypoint-initdb.d` folder.
```
Comment (sanikolaev) [inline README.md:339]:
```suggestion
To restore your full backup on startup, you need to mount your backup to the `/docker-entrypoint-initdb.d` folder. 
```

File: README.md [hunk 11]
```diff
@@ -312,6 +312,52 @@ There are several methods to build plain tables from your custom configuration f
...
+
+To restore your full backup, you need to mount your backup to the `/docker-entrypoint-initdb.d` folder. 
+Please note that you should mount the content of your backup, not the backup folder itself (e.g., `backup-202307..`).
+
+The restore process temporarily prevents the regular `searchd` start. 
+If you wish to avoid this behavior, you can set the flag `-e START_AFTER_RESTORE=true`.
+
+### Creating SQL dumps
+
+The simplest way to ensure this is to use `docker exec` and run the tool from the same container. Here's an example:
```
Comment (sanikolaev) [inline README.md:347]:
```suggestion
`manticore-backup` creates a physical backup. If you prefer a logical backup, you can use `mysqldump` in the container. For that use `docker exec` to log in to the container and run the tool. Here's an example:
```

File: README.md [hunk 12]
```diff
@@ -312,6 +312,52 @@ There are several methods to build plain tables from your custom configuration f
...
+
+```bash
+docker exec some-mysql sh -c 'exec mysqldump' > /some/path/on/your/host/dump.sql
+```
+
+### Restore SQL dumps
+
+For restoring data, you can use the `docker exec` command with the `-i` flag, similar to the following:
```
Comment (sanikolaev) [inline README.md:355]:
```suggestion
For restoring data from an sql file created by `mysqldump`, you can use the `docker exec` command with the `-i` flag like this:
```

**Review by sanikolaev — CHANGES_REQUESTED** [review]

File: docker-entrypoint.sh (hunk 1 above)
Comment (djklim87) [reply docker-entrypoint.sh:216]: It allows to check if the user mounted the backup path correctly (It's possible that the user will mount the folder instead backup's files)

File: docker-entrypoint.sh (hunk 2 above)
Comment (djklim87) [reply docker-entrypoint.sh]: done

File: docker-entrypoint.sh (hunk 1 above)
Comment (sanikolaev) [reply docker-entrypoint.sh:216]: Why 4? Looks like an error-prone approach that we'll have to deal with in the future.

**Pull request merged**